project_generator/project.py

```diff
@@ -0,0 +1,401 @@
+
+    """represents the settings that are specifc to targets"""
+
+    def __init__(self):
+        self.include_paths = []
+        self.source_paths = []
+        self.source_groups = {
+            'default': {}
+        }
+        self.macros = []
+        self.misc = {}
+
+        self.linker_file = None
+
+    def add_settings(self, data_dictionary, group_name):
+        if 'source_paths' in data_dictionary:
+            self.source_paths.extend(data_dictionary['source_paths'])
+
+        if 'source_files' in data_dictionary:
+            self._process_source_files(
+                data_dictionary['source_files'], group_name)
+
+        if 'include_paths' in data_dictionary:
+            self.include_paths += data_dictionary['include_paths']
+
+        if 'macros' in data_dictionary:
+            self.macros.extend(data_dictionary['macros'])
+
+        if 'project_dir' in data_dictionary:
+            self.project_dir.update(data_dictionary['project_dir'])
+
+        if 'linker_file' in data_dictionary:
+            self.linker_file = data_dictionary['linker_file'][0]
+
+        if 'misc' in data_dictionary:
+            self.misc.update(data_dictionary['misc'])
+
+    def source_of_type(self, filetype):
+        """return a dictionary of groups and the sources of a specified type within them"""
+        files = {}
+        for group, group_contents in self.source_groups.items():
+            files[group] = []
+            if filetype in group_contents:
+                files[group].extend(group_contents[filetype])
+
+        return files
+
+    def all_sources_of_type(self, filetype):
+        """return a list of the sources of a specified type"""
+        files = []
+
+        for group, group_contents in self.source_groups.items():
+            if filetype in group_contents:
+                files.extend(group_contents[filetype])
+
+        return files
+
+
+    def _process_source_files(self, files, group_name):
+        extensions = ['cpp', 'c', 's', 'obj', 'lib']
+
+        mappings = defaultdict(lambda: None)
+
+        mappings['o'] = 'obj'
+        mappings['a'] = 'lib'
+        mappings['ar'] = 'lib'
+        mappings['cc'] = 'cpp'
+
+        if group_name not in self.source_groups:
+            self.source_groups[group_name] = {}
+
+        for source_file in files:
+            extension = source_file.split('.')[-1]
+            extension = mappings[extension] or extension
+
+            if extension not in extensions:
+                continue
+
+            if extension not in self.source_groups[group_name]:
+                self.source_groups[group_name][extension] = []
+
+            self.source_groups[group_name][extension].append(source_file)
+
+            if os.path.dirname(source_file) not in self.source_paths:
+                self.source_paths.append(os.path.dirname(source_file))
+
+
+class Project:
+
+    """represents a project, which can be formed of many yaml files"""
+
+    TOOLCHAINS = {
+        'iar': 'iar',
```

Comment:
should be 'iar ewarm' or 'ewarm', to avoid confusion with other toolchains provided by iar